File: lib/domain/cultivation_plan/gateways/plan_save_user_pest_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      # 公開プラン保存: ユーザー害虫の検索・作成・関連付け（狭い永続化ポート）。
      class PlanSaveUserPestGateway
        # @return [Dtos::PlanSaveUserPestSnapshot, nil]
        def find_by_user_id_and_source_pest_id(user_id:, source_pest_id:)
          raise NotImplementedError
        end

        # @param user_id [Integer]
        # @param attributes [Hash]
        # @return [Dtos::PlanSaveUserPestSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordInvalid]
        def create(user_id:, attributes:)
          raise NotImplementedError
        end

        # @param pest_id [Integer]
        # @param attributes [Hash]
        def create_temperature_profile(pest_id:, attributes:)
          raise NotImplementedError
        end

        # @param pest_id [Integer]
        # @param attributes [Hash]
        def create_thermal_requirement(pest_id:, attributes:)
          raise NotImplementedError
        end

        # @param pest_id [Integer]
        # @param attributes [Hash]
        def create_control_method(pest_id:, attributes:)
          raise NotImplementedError
        end

        # @param crop_id [Integer]
        # @param pest_id [Integer]
        def link_crop_pest(crop_id:, pest_id:)
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_save_user_pest_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::PlanSaveUserPestGateway`

use crate::cultivation_plan::dtos::PlanSaveUserPestSnapshot;
use crate::shared::attr::AttrMap;

pub trait PlanSaveUserPestGateway: Send + Sync {
    fn find_by_user_id_and_source_pest_id(
        &self,
        user_id: i64,
        source_pest_id: i64,
    ) -> Result<Option<PlanSaveUserPestSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn create(
        &self,
        user_id: i64,
        attributes: AttrMap,
    ) -> Result<PlanSaveUserPestSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn create_temperature_profile(&self, pest_id: i64, attributes: AttrMap);

    fn create_thermal_requirement(&self, pest_id: i64, attributes: AttrMap);

    fn create_control_method(&self, pest_id: i64, attributes: AttrMap);

    fn link_crop_pest(&self, crop_id: i64, pest_id: i64);
}
